Source/DataSources/KmlDataSource.js

```diff
@@ -1474,27 +1628,370 @@ define([
+        var scalar = wMagnitude / qMagnitude;
+        var eastOffset = Cartesian3.multiplyByScalar(eUnit, scalar, scratchCartesian3_2);
+        var northOffset = Cartesian3.multiplyByScalar(nUnit, scalar, scratchCartesian3_3);
+
+        // A conservative measure for the longitudes would be to use the min/max longitudes of the bounding frustum.
+        var upperLeft = Cartesian3.add(center, northOffset, horizonPoints[0]);
+        Cartesian3.subtract(upperLeft, eastOffset, upperLeft);
+        Cartesian3.multiplyComponents(radii, upperLeft, upperLeft);
+
+        var lowerLeft = Cartesian3.subtract(center, northOffset, horizonPoints[1]);
+        Cartesian3.subtract(lowerLeft, eastOffset, lowerLeft);
+        Cartesian3.multiplyComponents(radii, lowerLeft, lowerLeft);
+
+        var upperRight = Cartesian3.add(center, northOffset, horizonPoints[2]);
+        Cartesian3.add(upperRight, eastOffset, upperRight);
+        Cartesian3.multiplyComponents(radii, upperRight, upperRight);
+
+        var lowerRight = Cartesian3.subtract(center, northOffset, horizonPoints[3]);
+        Cartesian3.add(lowerRight, eastOffset, lowerRight);
+        Cartesian3.multiplyComponents(radii, lowerRight, lowerRight);
+
+        return horizonPoints;
+    }
+
+
+    function computeViewRectangle(camera, canvas) {
+        if (!defined(camera) || !defined(canvas)) {
+            return Rectangle.MAX_VALUE;
+        }
+
+        var wgs84 = Ellipsoid.WGS84;
+
+        var width = canvas.clientWidth;
+        var height = canvas.clientHeight;
+
+        computeHorizonQuad(camera);
+
+        var result;
+        var count = 0;
+        function addToResult(x, y, index) {
+            scratchCartesian2_1.x = x;
+            scratchCartesian2_1.y = y;
+            var r = camera.pickEllipsoid(scratchCartesian2_1, wgs84, new Cartesian3());
+            if (defined(r)) {
+                wgs84.cartesianToCartographic(r, scratchCartographic);
+                ++count;
+            } else {
+                wgs84.cartesianToCartographic(horizonPoints[index], scratchCartographic);
+            }
+
+            if (!defined(result)) {
+                result = new Rectangle(scratchCartographic.longitude, scratchCartographic.latitude,
+                    scratchCartographic.longitude, scratchCartographic.latitude);
+            } else {
+                Rectangle.expand(result, scratchCartographic, result);
+            }
+        }
+
+        addToResult(0, 0, 0);
+        addToResult(0, height, 1);
+        addToResult(width, 0, 2);
+        addToResult(width, height, 3);
+
+        if (count === 0 || count === 1) {
+            // If we have space non-globe in 3 or 4 corners then retrn the whole globe
+            return Rectangle.MAX_VALUE;
+        }
+        
+        return result;
+    }
+
+    function makeQueryString() {
+        var result = '';
+        for(var i=0;i<arguments.length;++i) {
```

Comment:
So this combines query strings since we end up having to combine `NetworkLink.httpQuery`, `NetworkLink.viewFormat` and `NetworkLinkControl.cookie` into a single query string. It just makes sure none of them start with a `?` or `&`, so in the end we end up with a clean query string. It only gets called with 2 arguments, so I can just make it take 2 strings.